diffrax/_step_size_controller/jump_step_wrapper.py

```diff
@@ -0,0 +1,410 @@
+# infinity otherwise.
+
+# EXPLANATION OF REVISITING REJECTED STEPS
+# ----------------------------------------
+# We use a "stack" of rejected steps, composed of a buffer `rjct_buff` of length
+# `rejected_step_buffer_len` and a counter `i_rjct`. The "stack" are all the items
+# in `rjct_buff[i_rjct:]` with `rjct_buff[i_rjct]` being the top of the stack.
+# When `i_rjct == rejected_step_buffer_len`, the stack is empty.
+# At the start of the run, `i_rjct = rejected_step_buffer_len`. Each time a step is
+# rejected `i_rjct -=1` and `rjct_buff[i_rjct] = t1`. Each time a step ends at
+# `t1 == rjct_buff[i_rjct]`, we increment `i_rjct` by 1 (even if the step was
+# rejected, in which case we will re-add `t1` to the stack immediately).
+# We clip the next step to `t1_next = min(t1_next, rjct_buff[i_rjct])`.
+# If `i_rjct < 0` then an error is raised.
+
+
+class JumpStepWrapper(
+    AbstractStepSizeController[_JumpStepState[_ControllerState], _Dt0]
+):
+    """Wraps an existing step controller and adds the ability to specify `step_ts`
+    and `jump_ts`. The former are times to which the controller should step and the
+    latter are times at which the vector field has a discontinuity (jump)."""
+
+    controller: AbstractStepSizeController[_ControllerState, _Dt0]
+    step_ts: Optional[Real[Array, " steps"]]
+    jump_ts: Optional[Real[Array, " jumps"]]
+    rejected_step_buffer_len: int = eqx.field(static=True)
+    callback_on_reject: Optional[Callable] = eqx.field(static=True)
+
+    @eqxi.doc_remove_args("_callback_on_reject")
+    def __init__(
+        self,
+        controller,
+        step_ts=None,
+        jump_ts=None,
+        rejected_step_buffer_len=0,
+        _callback_on_reject=None,
+    ):
+        r"""
+        **Arguments**:
+
+        - `controller`: The controller to wrap.
+            Can be any diffrax.AbstractStepSizeController.
+        - `step_ts`: Denotes extra times that must be stepped to.
+        - `jump_ts`: Denotes extra times that must be stepped to, and at which the
+            vector field has a known discontinuity. (This is used to force FSAL solvers
+            so re-evaluate the vector field.)
+        - `rejected_step_buffer_len`: The length of the buffer storing rejected steps.
+            If this is > 0, then the controller will revisit rejected steps. This is
+            useful for SDEs, where the solution is guaranteed to be correct only if the
+            SDE is evaluated at all times at which the Brownian motion (BM) is
+            evaluated. Since the BM is also evaluated at rejected steps, we must later
+            evaluate the SDE at these times as well.
+        """
+        self.controller = controller
+        self.step_ts = _none_or_array(step_ts)
+        self.jump_ts = _none_or_array(jump_ts)
+        self.rejected_step_buffer_len = rejected_step_buffer_len
+        self.callback_on_reject = _callback_on_reject
+        self.__check_init__()
```

Comment:
This isn't necessary, Equinox will call this for you.

Reply:
Oh, thanks. This was actually added by copilot and I wasn't sure, so I left it 😅.